### ISSUE TITLE
Issue of reusing epoch_iter introduced from the latest commit

### ISSUE THREAD

**Issue by freewym** [pytorch/fairseq#1156]
@ngoyal2707 @myleott  At [this line](https://github.com/pytorch/fairseq/commit/32335404f09c47cccbfbf731abc4c510d0eef043#diff-a13cbd2246b6ad48cf7a6df9eb77be7eR131) from [this](https://github.com/pytorch/fairseq/commit/32335404f09c47cccbfbf731abc4c510d0eef043) commit, `self.epoch_iter` will not be updated if not None, i.e., the iterator would be the same across epochs. However, we need different seeds for the epoch iterator in different epochs for shuffling; otherwise we will get the same batch order across epochs.. So IMO returning same iterator across epochs might not be a good idea in any cases.

PS:  This also made the validation log printing starting from epoch 2. I think it is because the training iterator will even be reused for validation when reusing `self.epoch_iter` like that.

**Comment by ngoyal2707** [comment]
@freewym In fairseq, we first shuffle the samples, batch them by size for optimal padding and then finally shuffle the batches. So between two he epochs only batches are shuffled and the samples in the batches remain the same. This behavior was due to performance reasons. [In this commit](https://github.com/pytorch/fairseq/commit/32335404f09c47cccbfbf731abc4c510d0eef043), we made this configurable by task if the dataset itself is dynamic then we want to shuffle samples also every epoch.
Saying that, we are looking into making this default behavior and always shuffling samples also between epochs.

That's an issue with the validation iterator, looking at fixing it now.

**Comment by freewym** [comment]
@ngoyal2707 Yes. But the seed being passed to `EpochBatchIterator` seems to be used for shuffling batches, not samples. Shouldn't  shuffling batches be default for all tasks?

**Comment by ngoyal2707** [comment]
@freewym The batches are shuffled [here](https://github.com/pytorch/fairseq/blob/master/fairseq/data/iterators.py#L242) with `seed+epoch`, and the `shuffle` arg to [next_epoch_iter](https://github.com/pytorch/fairseq/blob/master/fairseq/data/iterators.py#L165) is set to `True` by default during training for all non-curriculum training [here](https://github.com/pytorch/fairseq/blob/master/train.py#L111). 

Let me know if this clears things up.

**Comment by freewym** [comment]
@ngoyal2707 Oh I see, you are right. I just want to confirm an experiment before/after this change still yield exactly the same result, as in my local experiment they are not. I am not if it is related to the validation iterator issue.

**Comment by ngoyal2707** [comment]
Sure, I just landed https://github.com/pytorch/fairseq/commit/3b09b98b662d4bbde50ebd067f7a14268b2eab1e , that should also fix your original issue of train epoch_iter being reused mistakenly for validation.

Let me know if you see any other issues.

**Comment by freewym** [comment]
@ngoyal2707 I pulled your latest commit, and it solved the validation issue, but it didn't give identical results as before. I wonder if skipping [this line](https://github.com/pytorch/fairseq/blob/master/fairseq/tasks/fairseq_task.py#L139) because of this [return](https://github.com/pytorch/fairseq/blob/master/fairseq/tasks/fairseq_task.py#L131) has made such inconsistency.

**Comment by freewym** [comment]
@ngoyal2707 OK, I find the reason why my previous experiment was not reproducible: [this commit](https://github.com/pytorch/fairseq/commit/10f9349e8a1f624255166aeef1c9c721de93041c) replaced `torch.rand()` with `torch.zeros()`, affecting the state of RNG when doing multi-gpu training. It may sound strange, but is it possible to revert back to torch.rand() so that all the previous experiments are reproducible?

**Comment by ngoyal2707** [comment]
yeah, the reason to change this was similar to what you mentioned as we were seeing slightly different results between single and multi gpu setup because of this extra `rand()` call in multi-gpu case.

I don't think the right solution here is to keep the `rand()` call. DO you see big difference in your experiments with and without it?
Worse case, I'd suggest add that `rand()` call in your task `__init__`?

(cc: @myleott )

**Comment by freewym** [comment]
@ngoyal2707 Good idea! Thanks!
